documentation/dataplane/dash-byo-data-plane-app.md

```diff
@@ -0,0 +1,142 @@
+# DASH BYO (Bring-Your-Own) data plane app
+
+1. [1. Data plane app](#1-data-plane-app)
+2. [2. BYO data plane app](#2-byo-data-plane-app)
+   1. [2.1. Work flow](#21-work-flow)
+   2. [2.2. DASH management role and ASIC programming](#22-dash-management-role-and-asic-programming)
+   3. [2.3. Flow management](#23-flow-management)
+   4. [2.4. RSS support](#24-rss-support)
+3. [3. SAI API design](#3-sai-api-design)
+   1. [3.1. SAI switch attribute updates](#31-sai-switch-attribute-updates)
+   2. [3.2. DASH pipeline programming APIs](#32-dash-pipeline-programming-apis)
+4. [4. Future work](#4-future-work)
+
+## 1. Data plane app
+
+Data plane app is one of the most important pieces in DASH pipeline. It is mainly responsible for:
+
+- Flow management, such as flow creation, deletion, resimulation, etc.
+- Work with ASIC to complete the table lookup and packet transformations. Depends on ASIC capability, if any operation could not be completed inside ASIC, it can be postponed to data plane app and completed in software.
+- ASIC management, such as initializing and programming match stage, etc.
+- and more.
+
+![DASH data plane app](./images/dash-default-data-plane-app.svg)
+
+By default, DASH technology providers will provide their own data plane app. This is usually closed source and can satisfy the DASH community requirement. However, in some cases, you might want to use your own data plane app due to some reasons, such as: private protocol support or support special logic in packet handling that is not supported by DASH technology providers. In this case, you can enable DASH BYO data plane app.
+
+## 2. BYO data plane app
+
+BYO data plane app is essentially a DPDK application that directly interacts with the data plane netdev.
+
+When BYO data plane app is enabled, the data plane app provided by technology providers will be either disabled or running without tied to the data plane netdev and only act as part of SAI API implementation if needed. With this setup, DASH users can start to use that device to run their own data plane app and process the packets.
+
+After the change, at a high level, the system architecture will be look like below:
+
+![DASH BYO data plane app](./images/dash-byo-data-plane-app.svg)
+
+### 2.1. Work flow
+
+The work flow of enabling the BYO data plane app will be like below:
+
+```mermaid
+sequenceDiagram
+
+participant User
+participant SAI
+participant IDPA as DASH Inbox data plane app
+participant BYODPA as BYO data plane app
+participant netdev
+
+User->>SAI: SAI create switch<br>with controller role<br>and settings
+SAI->>IDPA: configure inbox<br>data plane app
+User->>SAI: Get netdev name<br>as switch attribute
+User->>BYODPA: Launch and configure BYO data plane app
+BYODPA->>SAI: SAI create switch with worker role
+BYODPA->>netdev: Initialize on top of netdev
+```
+
+After initialization, the data plane app will be able to:
+
+- Receive/Send packets from/to netdev
+- Use SAI API to program the ASIC
```

Comment:
Would it be fair to say that using SAI APIs to program the ASIC would be optional? The existing SAI apis for flow management may or may not work for a custom data plane, so a BYO data plane may want to use low-level ASIC SDK APIs to manage ASIC tables.

Reply:
Hi Ashutosh, since the BYO data plane is essentially the customer's own implementation, so they could use the platform-dependent APIs or SAI for building the data plane. However, using SAI will give them the advantage of building platform independent applications.

Here, since the behavior of the platform-dependent APIs are not under control of SAI, so they are not included in the doc. And if there are more valid flow actions, we can also make SAI flow APIs supports it as well. The process should be similar as how SAI is working today. :D

Also sorry for the late reply. Just came back from vacation.

Reply:
I have updated the doc with more details and also have this ability noted in the doc too.

Reply:
What's the difference between this SAI api vs rte flow api used by DPDK app for hareware flow offloading?